src/smc-webapp/project/new/file-type-selector.tsx

```diff
@@ -0,0 +1,276 @@
+import * as React from "react";
+import { Col, Row } from "react-bootstrap";
+import { useSelector } from "react-redux";
+
+import { Tip } from "../../r_misc";
+
+import { JupyterServerPanel } from "../plain-jupyter-server";
+import { JupyterLabServerPanel } from "../jupyterlab-server";
+
+import { NewFileButton } from "./new-file-button";
+import { AvailableFeatures } from "./types";
+
+interface Props {
+  create_file: (name?: string) => void;
+  project_id?: string;
+  children?: React.ReactNode;
+  name: string;
+}
+
+// Use Rows and Cols to append more buttons to this component
+// Could be changed to auto adjust to a list of pre-defined button names.
+export function FileTypeSelector({
+  name,
+  create_file,
+  project_id,
+  children
+}: Props): JSX.Element | null {
+  const [show_jupyter_server, set_show_jupyter_server] = React.useState(false);
+  const [show_jupyterlab_server, set_show_jupyterlab_server] = React.useState(
+    false
+  );
+  // TODO: Add typechecking to this hook!
+  const available_features = useSelector<any, AvailableFeatures>(obj => {
+    return obj.getIn([name, "available_features"]);
+  });
+
+  if (!create_file || !create_file || !project_id) {
+    return null;
+  }
+  const row_style = { marginBottom: "8px" };
+  const available = available_features?.toJS() ?? {};
+
```

Comment:
Ugh, reading the code below I immediately want to replace all these 
```
<Tip... <NewFileButon...</Tip>
```
blocks by (1) putting all that data in a const object, then (2) a function that does all this.

And your rewrite of this makes it vastly safer and easier to make that sort of change later.  Yeah.  🤷‍♂